pkg/messaging/kafka/publisher.go

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package kafka
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/mainflux/mainflux/pkg/messaging"
+	"github.com/segmentio/kafka-go"
+)
+
+var (
+	_               messaging.Publisher = (*publisher)(nil)
+	backoffSchedule                     = []time.Duration{1 * time.Second, 3 * time.Second, 10 * time.Second}
+)
+
+type publisher struct {
+	conn *kafka.Conn
+	url  string
+}
+
+// NewPublisher returns Kafka message Publisher.
+func NewPublisher(url string) (messaging.Publisher, error) {
+	conn, err := kafka.Dial("tcp", url)
+	if err != nil {
+		return nil, err
+	}
+	ret := &publisher{
+		conn: conn,
+		url:  url,
+	}
+	return ret, nil
+
+}
+
+func (pub *publisher) Publish(topic string, msg messaging.Message) error {
+	if topic == "" {
+		return ErrEmptyTopic
+	}
+	data, err := proto.Marshal(&msg)
+	if err != nil {
+		return err
+	}
+	subject := fmt.Sprintf("%s.%s", chansPrefix, topic)
+	if msg.Subtopic != "" {
+		subject = fmt.Sprintf("%s.%s", subject, msg.Subtopic)
+	}
+	writer := kafka.NewWriter(kafka.WriterConfig{
```

Comment:
Check writer config options and try to find the best one for our use case. Looks like there are many options, backoff included (so you don't have to implement it here).

Reply:
They have implemented it but with no configurable parameters for the writer [here](https://github.com/mainflux/mainflux/pull/1579/files#r905379708)

Reply:
There is a `MaxAttempts` field that can be used for that. The lib is a little bit complex, but you can try and debug the code to check how it works.

Reply:
It backs off up to 1s then stops

```
kafka writer: writing 1 messages to channels.topic.engine (partition: 0)
kafka writer: error writing messages to channels.topic.engine (partition 0): kafka.(*Client).Produce: dial tcp 127.0.0.1:9092: connect: connection refused
kafka writer: backing off 100ms writing 1 messages to channels.topic.engine (partition: 0)
kafka writer: writing 1 messages to channels.topic.engine (partition: 0)
kafka writer: error writing messages to channels.topic.engine (partition 0): kafka.(*Client).Produce: dial tcp 127.0.0.1:9092: connect: connection refused
kafka writer: backing off 400ms writing 1 messages to channels.topic.engine (partition: 0)
kafka writer: writing 1 messages to channels.topic.engine (partition: 0)
kafka writer: error writing messages to channels.topic.engine (partition 0): kafka.(*Client).Produce: dial tcp 127.0.0.1:9092: connect: connection refused
kafka writer: backing off 900ms writing 1 messages to channels.topic.engine (partition: 0)
kafka writer: writing 1 messages to channels.topic.engine (partition: 0)
kafka writer: error writing messages to channels.topic.engine (partition 0): kafka.(*Client).Produce: dial tcp 127.0.0.1:9092: connect: connection refused
kafka writer: backing off 1s writing 1 messages to channels.topic.engine (partition: 0)
```

Reply:
That should be OK, 1s should be more than enough. In case of, say, 10 attempts - it's up to 10s. We assume that Kafka is ready to accept messages before we start and backoff is only for cases where something malfunctioning or Kafka is overloaded. How did you test this, did you wait some time after compose is up, and did you check Kafka logs to be sure it's up and ready to accept connections before you started testing?

Reply:
When testing it on the platform all other services throw this error message `{"level":"error","message":"Failed to connect to postgres: dial tcp 172.19.0.6:5432: connect: connection refused","ts":"2022-06-24T13:56:42.923294665Z"}`  but after that it connects. Publishing from one adapter and receiving in another works fine.